statefun-integration-tests/statefun-sanity-itcase/src/test/java/org/apache/flink/statefun/itcases/sanity/SanityVerificationITCase.java

```diff
@@ -0,0 +1,287 @@
+    }
+  }
+
+  private static Producer<FnAddress, Command> kafkaCommandProducer(String bootstrapServers) {
+    Properties props = new Properties();
+    props.put("bootstrap.servers", bootstrapServers);
+
+    return new KafkaProducer<>(
+        props, new FnAddressSerializerDeserializer(), new CommandSerializer());
+  }
+
+  private static Consumer<FnAddress, StateSnapshot> kafkaStateSnapshotConsumer(
+      String bootstrapServers) {
+    Properties consumerProps = new Properties();
+    consumerProps.setProperty("bootstrap.servers", bootstrapServers);
+    consumerProps.setProperty("group.id", "sanity-itcase");
+    consumerProps.setProperty("auto.offset.reset", "earliest");
+
+    KafkaConsumer<FnAddress, StateSnapshot> consumer =
+        new KafkaConsumer<>(
+            consumerProps, new FnAddressSerializerDeserializer(), new StateSnapshotDeserializer());
+    consumer.subscribe(Collections.singletonList(KafkaIO.STATE_SNAPSHOTS_TOPIC_NAME));
+
+    return consumer;
+  }
+
+  private static void produceCommandToKafka(
+      Producer<FnAddress, Command> producer, Command command) {
+    producer.send(new ProducerRecord<>(KafkaIO.COMMAND_TOPIC_NAME, command.getTarget(), command));
+  }
+
+  // =================================================================================
+  //  Protobuf message building utilities
+  // =================================================================================
+
+  private static StateSnapshot stateSnapshot(FnAddress fromFnAddress, int stateSnapshotValue) {
+    return StateSnapshot.newBuilder().setFrom(fromFnAddress).setState(stateSnapshotValue).build();
+  }
+
+  private static Command sendAction(FnAddress targetAddress, Command commandToSend) {
+    final Send sendAction = Send.newBuilder().addCommandToSend(commandToSend).build();
+
+    return Command.newBuilder().setTarget(targetAddress).setSend(sendAction).build();
+  }
+
+  private static Command modifyAction(FnAddress targetAddress, int stateValueDelta) {
+    final Modify modifyAction = Modify.newBuilder().setDelta(stateValueDelta).build();
+
+    return Command.newBuilder().setTarget(targetAddress).setModify(modifyAction).build();
+  }
+
+  private static Command noOpAction(FnAddress targetAddress) {
+    return Command.newBuilder().setTarget(targetAddress).setNoop(Noop.getDefaultInstance()).build();
+  }
+
+  private static FnAddress fnAddress(int typeIndex, String fnId) {
+    if (typeIndex > Constants.FUNCTION_TYPES.length - 1) {
+      throw new IndexOutOfBoundsException(
+          "Type index is out of bounds. Max index: " + (Constants.FUNCTION_TYPES.length - 1));
+    }
+    return FnAddress.newBuilder().setType(typeIndex).setId(fnId).build();
+  }
+
+  // =================================================================================
+  //  Kafka ingress / egress serde
+  // =================================================================================
+
+  public static final class FnAddressSerializerDeserializer
```

Comment:
Is it possible to replace all of these with a single `ProtobufSerializer<T extends Message>`,
not a big deal with it is not easy doable.

Reply:
Turns out there is a nice way to do that!
See fdcdef1c1a4697bbecb0d9949d002dbe76551f44 for the fix.